src/main.rs

```diff
@@ -253,8 +258,40 @@ impl EventHandler for Messages {
         self.cmds.execute(cx, msg);
     }
 
+    fn message_update(
+        &self,
+        cx: Context,
+        _: Option<Message>,
+        _: Option<Message>,
+        event: MessageUpdateEvent,
+    ) {
+        use serenity::utils::CustomMessage;
+        let mut msg = CustomMessage::new();
+
+        msg.id(event.id)
+            .channel_id(event.channel_id)
+            .content(event.content.unwrap_or_else(|| String::new()));
+
+        let msg = msg.build();
+        self.cmds.execute(cx, msg);
+    }
+
+    fn message_delete(&self, ctx: Context, channel_id: ChannelId, deleted_id: MessageId) {
+        let mut data = ctx.data.write();
+        let history = data.get_mut::<CommandHistory>().unwrap();
+        if let Some(response_id) = history.remove(&deleted_id) {
+            info!("deleting message: {:?}", response_id);
+            let _ = channel_id.delete_message(&ctx, response_id);
+        }
+    }
+
     fn ready(&self, context: Context, ready: Ready) {
         info!("{} connected to discord", ready.user.name);
+
+        let mut data = context.data.write();
+        data.insert::<CommandHistory>(HashMap::new());
+        drop(data);
```

Comment:
Why are we explicitly dropping this here?

Reply:
if we don't drop the guard we can't move context into the unban thread, since the borrow lasts to the end of the block

(we could clone the context, its just a bunch of Arcs)